File: movie-manager/app/pages/login-page/login-view-model.ts
import { Observable } from "data/observable";
import { EventData } from 'ui/text-base';
import * as loginService from '../../services/login-service';
import * as navigationModule from '../../shared/navigation';

export class LoginViewModel extends Observable {
    private _username: string;
    private _password: string;
    private _loginError: boolean;

    get loginError(): boolean {
        return this._loginError;
    }

    set loginError(value: boolean) {
        if (value !== this._loginError) {
            this._loginError = value;
            this.notifyPropertyChange('loginError', value);
        }
    }

    get username(): string {
        return this._username;
    }

    set username(value: string) {
        if (value !== this._username) {
            this._username = value;
            this.notifyPropertyChange('username', value);
        }
    }

    get password(): string {
        return this._password;
    }

    set password(value: string) {
        if (value !== this._password) {
            this._password = value;
            this.notifyPropertyChange('password', value);
        }
    }

    public constructor() {
        super();
        this.loginError = false;
        let savedCredentials = loginService.getSavedCredentials();
        if (savedCredentials) {
            this.username = savedCredentials.username;
            this.password = savedCredentials.password;
        }
    }

    processLogin(args: EventData) {
        this.loginError = false;
        loginService.processLogin(this.username, this.password).then(loginResponse => {
            if (loginResponse.success) {
                console.log('login success');
                loginService.addCredentials({username: this.username, password: this.password, userId: loginResponse.userId});
                navigationModule.navigateToMainPage(loginResponse.userId);
            } else {
                this.loginError = true;
                this.password = '';
                this.username = '';
                console.log('login failed');
            }
        });
    }
}

File: MovieManagerXamarin2/MovieManagerXamarin2/MovieManagerXamarin2/ViewModels/LoginViewModel.cs
using System.Windows.Input;
using MovieManagerXamarin2.Pages;
using MovieManagerXamarin2.Services;
using Xamarin.Forms;

namespace MovieManagerXamarin2.ViewModels
{
    public class LoginViewModel : BaseViewModel
    {
        private ILoginService _loginService;
        private string _username;
        private string _password;
        private bool _isReady;

        public string Username
        {
            get => _username;
            set => SetProperty(ref _username, value, "Username");
        }

        public string Password
        {
            get => _password;
            set => SetProperty(ref _password, value, "Password");
        }

        public bool IsReady
        {
            get => _isReady;
            set
            {
                if (value != _isReady)
                {
                    _isReady = value;
                    OnPropertyChanged(nameof(IsReady));
                    OnPropertyChanged(nameof(IsPaused));
                }
            }
        }

        public bool IsPaused
        {
            get => !_isReady;
        }

        public ICommand LoginTap { get; }

        public LoginViewModel()
        {
            _loginService = new LoginService();

            var credentials = _loginService.GetSavedCredentials();
            _username = credentials?.Username;
            _password = credentials?.Password;

            //LoginTap = new Command(async () =>
            //{
            //    var waitViewModel = new WaitViewModel(ProcessLogin);
            //    await Navigation.PushAsync(new WaitingPage {BindingContext = waitViewModel});
            //    waitViewModel.ProcessWait();
            //});

            LoginTap = new Command(ProcessLogin);

            _isReady = true;
        }

        protected async void ProcessLogin()
        {
            IsReady = false;
            var loginResponse = await _loginService.Login(Username, Password);
            if (loginResponse.Success)
            {
                var mainViewModel = new MainViewModel(loginResponse.UserId)
                {
                    Navigation = Navigation
                };

                await _loginService.SaveCredentials(Username, Password, loginResponse.UserId);
                await App.GetInstance().NavigationService.BackToMainPage();
                IsReady = true;
            }
        }
    }
}
